Modernize Library

patricksanders opened the pull request:
It seems that this repo has all but been abandoned, so I figured I would give it some love.

This is a wholesale rewrite of the MARTA API Python library. It implements two API wrapper functions, `get_buses()` and `get_trains()`. These functions take various keyword arguments to filter results and return lists of `Bus` or `Train` objects.

The design goal of these changes is to simplify API integration to allow for easy development using the MARTA API. I also wanted to lay the groundwork for extensibility and testability of the marta-python package.

Commit:
starting to build framework for api wrapper

Commit:
implement basic bus and train filtering

Commit:
add some basic unit tests

Commit:
add skeleton of setup.py

Commit:
remove old implementation, add more complete gitignore

Commit:
remove json file

Commit:
simplify vehicle classes

Commit:
cache api responses

Commit:
make line filter case insensitive

Commit:
iterate over trains fewer times

Commit:
cleanup, documentation, and basic error checking

Commit:
mock API request, add tests for vehicle creation

Commit:
update requirements

gwengullett commented on tests/test_api.py:
```diff
@@ -0,0 +1,82 @@
+import pytest
+import requests_mock
```
This import is not set in tests_require in setup.py nor requirements.txt

patricksanders replied:
Good catch. Just switched to tox to manage testing, so all should be good now. 👍

Commit:
use tox to manage testing

Commit:
Try/catching existing, but bad, API keys (#1)

* Removed extraneous slash from API calls

Because it's about the little things in life.

* Added basic try/catch on bad API key in get_trains

It seems like you don't actually need an API key for get_buses, which is weird, but if you request get_trains with an invalid API key MARTA returns an unparseable error message that you can only see in your browser.

* Check against response.status_code instead of loads() fail

patricksanders commented:
@landonreed Any chance you could review this? I'm also willing to maintain this repo if you've got too much on your plate.

landonreed commented on marta/api.py:
```diff
@@ -0,0 +1,76 @@
+    else:
+        url = '{}{}?apikey={}'.format(_BASE_URL, _BUS_PATH, _API_KEY)
+
+    response = requests.get(url)
```
Any reason why this response isn't checked for `401` or `403` status codes (as in `get_trains`)?

patricksanders replied:
AFAIK the bus endpoint doesn't actually require an API key. I added a check in case that changes.

landonreed commented on marta/api.py:
```diff
@@ -0,0 +1,76 @@
+    data = loads(response.text)
+    trains = [Train(t) for t in data]
+
+    trains = [t for t in trains if
```
This could probably use a comment summarizing the filter behaviors.

patricksanders replied:
Done. Also added a `TODO` to make that less terrible to look at.

landonreed reviewed (changes requested):
Looks great, sorry for the delay there!  I just had a couple of comments.

patricksanders commented:
Thanks @landonreed! I'll address your comments as soon as I get a chance.

landonreed commented on README.md:
````diff
@@ -1,7 +1,112 @@
+## Installing
+
+```
+python setup.py install
````
I'm not really a python developer (so I don't know all of the conventions), but should the README mention that you need python 3? I haven't run python code in a while and got stuck on an error importing `JSONDecodeError` when using 2.7

patricksanders replied:
This library should be compatible with 2.7 and 3.x. `JSONDecodeError` wasn't used anyway, so I removed it and the tests pass against 2.7 and 3.7. Nice catch!

Commit:
remove unused import, check for auth on buses

Commit:
bump tox to python 3.7

Commit:
add comment for unintuitive list comprehension

Commit:
add flake8 config, fix stuff that it found

landonreed commented:
Great stuff, thanks for the changes, @patricksanders!
